LANGUAGE.md

```diff
@@ -0,0 +1,155 @@
+The program is represented by a transition system that specifies how the state
+evolves in time in the form of constraints.
+
+As an example, let's apply that model conceptually to a program that computes
+the factorial of a number. Our state consists of two registers:
+
+- CNT - a counter
+- ACC - the answer accumulator
+
+The first row will represent the initial state of our program, that is, `PC =
+0`, `CNT = <input>`, `ACC = 1`.
+From then on, we compute the subsequent rows with the following pseudocode:
+
+1. START:
+2. Is CNT = 0?
+3. If `yes` go to `END`
+4. ACC := ACC * CNT
+5. CNT := CNT - 1
+6. Go to START
+7. END:
+8. ACC equals `factorial(input)`
+
+We can compute the following execution trace for the specific input 5:
+
+| CNT | ACC |
+|-----|-----|
+| 5   | 1   |
+| 4   | 5   |
+| 3   | 20  |
+| 2   | 60  |
+| 1   | 120 |
+| 0   | **120** |
+
+Note that the size of the table determines how long the trace can be.
+
+## ISA
+
+There are three constructs that can be used to define an instruction set:
+
+- Registers
+- Macros
+- Instructions
+
+### Registers
+
+Registers are basic state components that can store a single numbers element
+each. There are three types of registers:
+
+- Program Counter (PC). Used for control flow.
+- Assignment registers. Used by parametric instructions.
+- General registers. Used by program logic.
+
+The registers above can be declared in the following ways:
+
+```asm
+reg pc[@pc];
+reg X[<=];
+reg A;
+```
+
+In the snippet above, the `@pc` annotation means that `pc` is implicitly
+incremented by one for instructions that do not use it.  The `<=` annotation
+means `X` is an assignment register and can be used by parametric instructions.
+Finally, `A` is a general purpose register and can be used by program logic.
+
+### Instructions
+
+Instructions are predicates that define constraints that must hold when the
+instruction is invoked. The constraints may refer to registers on the execution
+trace line that it is invoked and/or the next line.
+
+```asm
+instr ADD X, Y -> Z { X + Y = Z }
+```
+
+The above instruction can be read as "If `Z` is the same as `X + Y`, then `ADD
+X Y` returns `Z`".
+As an analogy, this is similar to Prolog predicates.
+
+
+### Macros
+
+Macros can be used to build expressions which are inlined at the call site.
+
+```asm
+macro if_then_else(condition, true_value, false_value) {
+    condition * true_value + (1 - condition) * false_value
+};
+```
+
+The macro above can be used to simplify if-then-else arithmetic expressions.
+
+### Factorial, again
+
+We now have enough constructs to create a simple ISA and implement our
+Factorial program.
+
+```asm
+pil {
+    col witness XInv;
+    col witness XIsZero;
+    XIsZero * (1 - XIsZero) = 0;
+    XIsZero = 1 - X * XInv;
+    XIsZero * X = 0;
+}
+
+macro if_then_else(condition, true_value, false_value) {
+    condition * true_value + (1 - condition) * false_value
+};
+
+macro jump_to(target) { pc' = target; };
```

Comment:
Now the big question: Do we require `pc` to be a parameter?

Reply:
Yes please

Reply:
I'll leave it like this in this file for now